### DISCUSSION TITLE
Simulation Study: Introducing FasttextSum

### DISCUSSION THREAD

**Discussion by rohitgarud** [asreview/asreview#1420] [Show and tell]
Hello everyone,
Here I am introducing [ASReview-FasttextSum](https://github.com/rohitgarud/Asreview-FasttextSum), a new feature extraction technique (new in ASReview, not novel) based on the weighted sum of FastText word embeddings for generating the feature vectors (document embeddings). There are five different options of weighting (pooling) available: mean, TFIDF (Term Frequency Inverse Document Frequency), TF (Tem Frequency), Log base 2 of TF and [Smoothed Inverse Frequency (SIF)](https://openreview.net/pdf?id=SyK00v5xx). The stopwords can also be removed. 

SIF pooling performs well in the case of most of the datasets. The performance in terms of ATD can be improved by increasing the number of epochs. The presented results are for only 33 epochs of Fasttext model training. The stopwords were kept during this simulation study.

**Comment by rohitgarud** [comment]
| Dataset                 | Classifier          | Features    | Pooling | recall_0.5 | wss_0.95 | erf_0.1 | atd      |
|-------------------------|---------------------|-------------|---------|------------|----------|---------|----------|
| ACEInhibitors           | Naïve Bayes         | TFIDF       |         | 0.950      | 0.782    | 0.750   | 190.325  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.950      | 0.576    | 0.625   | 312.275  |
|                         |                     |             | TFIDF   | 0.925      | 0.427    | 0.575   | 376.350  |
|                         |                     |             | TF      | 0.825      | 0.030    | 0.350   | 668.175  |
|                         |                     |             | Log2_TF | 0.900      | 0.082    | 0.475   | 484.675  |
|                         |                     |             | SIF     | 0.975      | 0.719    | 0.650   | 206.925  |
| ADHD                    | Naïve Bayes         | TFIDF       |         | 0.947      | 0.446    | 0.737   | 96.789   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.895      | 0.223    | 0.579   | 138.632  |
|                         |                     |             | TFIDF   | 0.895      | 0.352    | 0.579   | 128.421  |
|                         |                     |             | TF      | 0.737      | 0.150    | 0.421   | 203.684  |
|                         |                     |             | Log2_TF | 0.895      | 0.344    | 0.526   | 137.316  |
|                         |                     |             | SIF     | 0.947      | 0.477    | 0.579   | 98.947   |
| Antihistamines          | Naïve Bayes         | TFIDF       |         | 0.867      | 0.091    | 0.267   | 72.333   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.800      | 0.250    | 0.267   | 84.267   |
|                         |                     |             | TFIDF   | 0.867      | 0.234    | 0.333   | 77.600   |
|                         |                     |             | TF      | 0.867      | 0.308    | 0.267   | 76.933   |
|                         |                     |             | Log2_TF | 0.800      | 0.214    | 0.400   | 75.333   |
|                         |                     |             | SIF     | 0.733      | 0.110    | 0.133   | 94.800   |
| Appenzeller-Herzog_2020 | Naïve Bayes         | TFIDF       |         | 0.964      | 0.723    | 0.714   | 263.464  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 1.000      | 0.716    | 0.750   | 274.214  |
|                         |                     |             | TFIDF   | 0.964      | 0.685    | 0.571   | 352.536  |
|                         |                     |             | TF      | 0.964      | 0.592    | 0.214   | 531.393  |
|                         |                     |             | Log2_TF | 0.964      | 0.682    | 0.750   | 280.643  |
|                         |                     |             | SIF     | 0.964      | 0.638    | 0.750   | 319.536  |
| AtypicalAntipsychotics  | Naïve Bayes         | TFIDF       |         | 0.869      | 0.168    | 0.255   | 275.124  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.772      | 0.163    | 0.234   | 315.703  |
|                         |                     |             | TFIDF   | 0.703      | 0.095    | 0.207   | 372.807  |
|                         |                     |             | TF      | 0.676      | 0.075    | 0.145   | 415.076  |
|                         |                     |             | Log2_TF | 0.710      | 0.165    | 0.200   | 353.048  |
|                         |                     |             | SIF     | 0.772      | 0.152    | 0.179   | 344.221  |
| Bannach-Brown_2019      | Naïve Bayes         | TFIDF       |         | 0.932      | 0.404    | 0.423   | 293.022  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.867      | 0.146    | 0.373   | 430.817  |
|                         |                     |             | TFIDF   | 0.803      | 0.127    | 0.229   | 539.090  |
|                         |                     |             | TF      | 0.638      | 0.027    | 0.129   | 789.946  |
|                         |                     |             | Log2_TF | 0.738      | 0.036    | 0.251   | 619.018  |
|                         |                     |             | SIF     | 0.968      | 0.558    | 0.384   | 278.190  |
| BetaBlockers            | Naïve Bayes         | TFIDF       |         | 0.927      | 0.669    | 0.610   | 241.878  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.878      | 0.267    | 0.634   | 302.439  |
|                         |                     |             | TFIDF   | 0.878      | 0.306    | 0.488   | 332.341  |
|                         |                     |             | TF      | 0.829      | 0.209    | 0.366   | 458.659  |
|                         |                     |             | Log2_TF | 0.854      | 0.206    | 0.512   | 341.927  |
|                         |                     |             | SIF     | 0.927      | 0.563    | 0.537   | 272.854  |
| Bos_2018                | Naïve Bayes         | TFIDF       |         | 1.000      | 0.834    | 0.900   | 76.000   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 1.000      | 0.753    | 0.800   | 327.400  |
|                         |                     |             | TFIDF   | 1.000      | 0.735    | 0.700   | 343.000  |
|                         |                     |             | TF      | 0.800      | 0.319    | 0.500   | 885.000  |
|                         |                     |             | Log2_TF | 1.000      | 0.689    | 0.600   | 440.400  |
|                         |                     |             | SIF     | 1.000      | 0.780    | 0.800   | 219.100  |
| CalciumChannelBlockers  | Naïve Bayes         | TFIDF       |         | 0.899      | 0.229    | 0.172   | 268.364  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.646      | 0.101    | 0.081   | 489.444  |
|                         |                     |             | TFIDF   | 0.646      | 0.072    | 0.101   | 482.485  |
|                         |                     |             | TF      | 0.535      | 0.109    | 0.030   | 535.727  |
|                         |                     |             | Log2_TF | 0.616      | 0.097    | 0.081   | 495.293  |
|                         |                     |             | SIF     | 0.758      | 0.188    | 0.030   | 449.818  |
| Estrogens               | Naïve Bayes         | TFIDF       |         | 0.911      | 0.249    | 0.177   | 89.418   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.658      | 0.098    | 0.025   | 149.709  |
|                         |                     |             | TFIDF   | 0.671      | 0.038    | 0.038   | 154.051  |
|                         |                     |             | TF      | 0.570      | 0.014    | -0.013  | 173.000  |
|                         |                     |             | Log2_TF | 0.646      | 0.036    | 0.000   | 155.051  |
|                         |                     |             | SIF     | 0.747      | 0.284    | 0.038   | 131.620  |
| Hall_2012               | Naïve Bayes         | TFIDF       |         | 1.000      | 0.911    | 0.893   | 134.272  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 1.000      | 0.765    | 0.767   | 435.136  |
|                         |                     |             | TFIDF   | 0.971      | 0.687    | 0.767   | 516.728  |
|                         |                     |             | TF      | 0.709      | 0.155    | 0.252   | 2895.728 |
|                         |                     |             | Log2_TF | 0.913      | 0.363    | 0.330   | 1494.291 |
|                         |                     |             | SIF     | 1.000      | 0.874    | 0.903   | 213.350  |
| Kitchenham_2010         | Naïve Bayes         | TFIDF       |         | 0.977      | 0.652    | 0.568   | 200.136  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.909      | 0.373    | 0.318   | 298.159  |
|                         |                     |             | TFIDF   | 0.773      | 0.239    | 0.205   | 477.841  |
|                         |                     |             | TF      | 0.682      | 0.001    | 0.000   | 668.318  |
|                         |                     |             | Log2_TF | 0.773      | 0.126    | 0.159   | 492.455  |
|                         |                     |             | SIF     | 1.000      | 0.633    | 0.636   | 160.545  |
| Kwok_2020               | Naïve Bayes         | TFIDF       |         | 0.992      | 0.675    | 0.588   | 227.092  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.899      | 0.244    | 0.319   | 512.588  |
|                         |                     |             | TFIDF   | 0.857      | 0.323    | 0.311   | 585.748  |
|                         |                     |             | TF      | 0.647      | 0.047    | 0.092   | 955.429  |
|                         |                     |             | Log2_TF | 0.798      | 0.111    | 0.185   | 728.790  |
|                         |                     |             | SIF     | 0.983      | 0.618    | 0.303   | 329.504  |
| Nagtegaal_2019          | Naïve Bayes         | TFIDF       |         | 0.990      | 0.529    | 0.540   | 237.190  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.930      | 0.392    | 0.390   | 324.720  |
|                         |                     |             | TFIDF   | 0.880      | 0.208    | 0.260   | 473.650  |
|                         |                     |             | TF      | 0.640      | 0.067    | 0.080   | 742.420  |
|                         |                     |             | Log2_TF | 0.860      | 0.187    | 0.230   | 482.300  |
|                         |                     |             | SIF     | 1.000      | 0.608    | 0.550   | 200.160  |
| NSAIDS                  | Naïve Bayes         | TFIDF       |         | 0.975      | 0.698    | 0.500   | 46.775   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.975      | 0.657    | 0.500   | 49.150   |
|                         |                     |             | TFIDF   | 0.975      | 0.621    | 0.550   | 49.975   |
|                         |                     |             | TF      | 0.950      | 0.529    | 0.425   | 60.975   |
|                         |                     |             | Log2_TF | 0.975      | 0.614    | 0.425   | 53.175   |
|                         |                     |             | SIF     | 0.975      | 0.655    | 0.425   | 47.075   |
| Opiods                  | Naïve Bayes         | TFIDF       |         | 0.929      | 0.467    | -0.071  | 717.857  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.857      | 0.194    | -0.071  | 596.429  |
|                         |                     |             | TFIDF   | 0.929      | 0.429    | -0.071  | 543.643  |
|                         |                     |             | TF      | 0.786      | 0.126    | -0.071  | 706.643  |
|                         |                     |             | Log2_TF | 0.857      | 0.369    | -0.071  | 583.929  |
|                         |                     |             | SIF     | 0.929      | 0.458    | -0.071  | 533.214  |
| OralHypoglycemics       | Naïve Bayes         | TFIDF       |         | 0.770      | 0.190    | 0.141   | 155.444  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.652      | 0.040    | 0.044   | 200.607  |
|                         |                     |             | TFIDF   | 0.681      | 0.066    | 0.037   | 194.311  |
|                         |                     |             | TF      | 0.541      | 0.072    | -0.030  | 240.037  |
|                         |                     |             | Log2_TF | 0.652      | 0.044    | 0.030   | 210.170  |
|                         |                     |             | SIF     | 0.681      | 0.032    | 0.052   | 196.711  |
| ProtonPumpInhibitors    | Naïve Bayes         | TFIDF       |         | 0.940      | 0.481    | 0.560   | 191.960  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.900      | 0.352    | 0.600   | 197.440  |
|                         |                     |             | TFIDF   | 0.900      | 0.376    | 0.500   | 226.620  |
|                         |                     |             | TF      | 0.720      | 0.221    | 0.260   | 383.840  |
|                         |                     |             | Log2_TF | 0.880      | 0.387    | 0.480   | 240.660  |
|                         |                     |             | SIF     | 0.920      | 0.342    | 0.540   | 190.540  |
| Radjenovic_2013         | Naïve Bayes         | TFIDF       |         | 1.000      | 0.841    | 0.851   | 189.574  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.957      | 0.618    | 0.660   | 607.426  |
|                         |                     |             | TFIDF   | 0.957      | 0.469    | 0.660   | 671.362  |
|                         |                     |             | TF      | 0.766      | 0.047    | 0.255   | 1874.468 |
|                         |                     |             | Log2_TF | 0.894      | 0.157    | 0.447   | 1061.255 |
|                         |                     |             | SIF     | 0.979      | 0.851    | 0.830   | 213.979  |
| SkeletalMuscleRelaxants | Naïve Bayes         | TFIDF       |         | 0.625      | -0.015   | -0.125  | 782.625  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.750      | -0.162   | -0.125  | 720.250  |
|                         |                     |             | TFIDF   | 0.500      | -0.170   | -0.125  | 1061.125 |
|                         |                     |             | TF      | 0.625      | -0.165   | -0.125  | 946.875  |
|                         |                     |             | Log2_TF | 0.625      | -0.160   | -0.125  | 808.375  |
|                         |                     |             | SIF     | 0.625      | 0.265    | -0.125  | 712.625  |
| Statins                 | Naïve Bayes         | TFIDF       |         | 0.940      | 0.471    | 0.631   | 445.036  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.821      | 0.090    | 0.393   | 718.869  |
|                         |                     |             | TFIDF   | 0.857      | 0.100    | 0.333   | 843.762  |
|                         |                     |             | TF      | 0.655      | 0.120    | 0.143   | 1287.619 |
|                         |                     |             | Log2_TF | 0.821      | 0.111    | 0.357   | 891.893  |
|                         |                     |             | SIF     | 0.821      | 0.069    | 0.548   | 679.333  |
| Triptans                | Naïve Bayes         | TFIDF       |         | 0.957      | 0.520    | 0.478   | 98.565   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.913      | 0.414    | 0.391   | 128.261  |
|                         |                     |             | TFIDF   | 0.870      | 0.167    | 0.391   | 135.565  |
|                         |                     |             | TF      | 0.870      | 0.381    | 0.348   | 160.391  |
|                         |                     |             | Log2_TF | 0.913      | 0.448    | 0.000   | 166.739  |
|                         |                     |             | SIF     | 0.783      | 0.286    | 0.348   | 141.913  |
| UrinaryIncontinence     | Naïve Bayes         | TFIDF       |         | 0.872      | 0.382    | 0.359   | 60.923   |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.974      | 0.517    | 0.385   | 52.744   |
|                         |                     |             | TFIDF   | 0.923      | 0.434    | 0.385   | 52.923   |
|                         |                     |             | TF      | 0.949      | 0.486    | 0.359   | 57.410   |
|                         |                     |             | Log2_TF | 0.974      | 0.489    | 0.359   | 53.103   |
|                         |                     |             | SIF     | 0.923      | 0.403    | 0.359   | 55.564   |
| van_de_Schoot_2017      | Naïve Bayes         | TFIDF       |         | 1.000      | 0.888    | 0.905   | 101.571  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.976      | 0.701    | 0.738   | 405.071  |
|                         |                     |             | TFIDF   | 0.929      | 0.613    | 0.524   | 703.143  |
|                         |                     |             | TF      | 0.714      | 0.140    | 0.238   | 1952.095 |
|                         |                     |             | Log2_TF | 0.952      | 0.484    | 0.476   | 911.476  |
|                         |                     |             | SIF     | 1.000      | 0.862    | 0.857   | 167.524  |
| van_Dis_2020            | Naïve Bayes         | TFIDF       |         | 0.972      | 0.650    | 0.556   | 1191.264 |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.903      | 0.236    | 0.236   | 2462.736 |
|                         |                     |             | TFIDF   | 0.861      | 0.292    | 0.125   | 2912.111 |
|                         |                     |             | TF      | 0.583      | 0.050    | 0.181   | 4374.139 |
|                         |                     |             | Log2_TF | 0.833      | 0.196    | 0.181   | 3250.208 |
|                         |                     |             | SIF     | 1.000      | 0.708    | 0.403   | 1149.583 |
| Wahono_2015             | Naïve Bayes         | TFIDF       |         | 1.000      | 0.863    | 0.869   | 197.803  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 0.984      | 0.663    | 0.574   | 656.262  |
|                         |                     |             | TFIDF   | 0.967      | 0.565    | 0.656   | 652.082  |
|                         |                     |             | TF      | 0.557      | 0.078    | 0.049   | 3034.230 |
|                         |                     |             | Log2_TF | 0.803      | 0.231    | 0.344   | 1695.918 |
|                         |                     |             | SIF     | 1.000      | 0.854    | 0.885   | 160.328  |
| Wolters_2018            | Naïve Bayes         | TFIDF       |         | 1.000      | 0.794    | 0.778   | 162.222  |
|                         | Similarity (Cosine) | FasttextSum | Mean    | 1.000      | 0.709    | 0.222   | 642.222  |
|                         |                     |             | TFIDF   | 0.944      | 0.456    | 0.222   | 980.500  |
|                         |                     |             | TF      | 0.444      | 0.208    | 0.000   | 2330.833 |
|                         |                     |             | Log2_TF | 1.000      | 0.555    | 0.111   | 1083.778 |
|                         |                     |             | SIF     | 1.000      | 0.817    | 0.833   | 197.833  |

**Reply by Rensvandeschoot** [reply]
Great simulation study (again!!). For what dataset properties would you recommend using Fasttext?

**Reply by rohitgarud** [reply]
Thank you @Rensvandeschoot. Currently, I am assessing whether training the FastText model for longer can improve the performance in terms of ATD (as you mentioned). Then I will try to correlate different aspects of the dataset with the performance such as the number of records in the dataset, the ratio of relevant/irrelevant, the topic or domain of the dataset, etc. Please let me know if you have any suggestions.

**Reply by rohitgarud** [reply]
Average for different options. SIF performs the best, will be testing it further.

| Classifier          | Features    | Pooling | recall_0.5 | wss_0.95 | erf_0.1 | atd      |
|---------------------|-------------|---------|------------|----------|---------|----------|
| Naïve Bayes         | TFIDF       |         | 0.934      | 0.540    | 0.519   | 259.520  |
| Similarity (Cosine) | FasttextSum | Mean    | 0.884      | 0.374    | 0.397   | 438.258  |
|                     |             | TFIDF   | 0.857      | 0.331    | 0.354   | 527.399  |
|                     |             | TF      | 0.714      | 0.163    | 0.180   | 1015.150 |
|                     |             | Log2_TF | 0.839      | 0.269    | 0.286   | 651.527  |
|                     |             | SIF     | 0.904      | 0.511    | 0.458   | 287.622  |

**Reply by rohitgarud** [reply]
Effect of the training epochs on the performance

| Classifier  | Features          | Training Epochs | recall_0.5 | wss_0.95 | erf_0.1 | atd     |
|-------------|-------------------|-----------------|------------|----------|---------|---------|
| Naïve Bayes | TFIDF             |                 | 0.934      | 0.540    | 0.519   | 259.520 |
|Similarity (Cosine)| FasttextSum (SIF) | 33              | 0.904      | 0.511    | 0.458   | 287.622 |
|             |                   | 100             | 0.920      | 0.525    | 0.473   | 279.318 |
|             |                   | 200             | 0.913      | 0.531    | 0.481   | 281.792 |
|             |                   | 300             | 0.924      | 0.529    | 0.495   | 279.980 |
|             |                   | 400             | 0.919      | 0.520    | 0.484   | 286.036 |

**Reply by Rensvandeschoot** [reply]
smaller effect than I expected.. interesting!

**Reply by rohitgarud** [reply]
Yes.. I was also expecting better performance.. I think still 100 epochs can be selected as default.. takes longer for bigger datasets.. should I post the detailed results?

**Reply by rohitgarud** [reply]
Also, there are a lot more parameters in FastText itself, tuning those might improve the performance. Need extensive experimentation for finding optimum parameters. The feature vector size used here is 40 which is another hyperparameter that needs to be tuned.
